src/app/composer/qgscompositionwidget.cpp

```diff
@@ -53,12 +52,8 @@ QgsPanelWidget( parent )
   updateVariables();
   connect( mVariableEditor, SIGNAL( scopeChanged() ), this, SLOT( variablesChanged() ) );
   // listen out for variable edits
-  QgsApplication* app = qobject_cast<QgsApplication*>( QgsApplication::instance() );
-  if ( app )
-  {
-    connect( app, SIGNAL( settingsChanged() ), this, SLOT( updateVariables() ) );
-  }
-  connect( QgsProject::instance(), SIGNAL( variablesChanged() ), this, SLOT( updateVariables() ) );
+  connect( QgsApplication::instance(), &QgsApplication::customVariablesChanged, this, &QgsCompositionWidget::updateVariables );
```

Comment:
I don't like it how the naming changes between application and project - can we standardise to either just variables or customVariables? (ExpressionVaraibles might be another more descriptive option)